scipy/special/_logsumexp.py

```diff
@@ -100,53 +103,117 @@ def logsumexp(a, axis=None, b=None, keepdims=False, return_sign=False):
     """
     xp = array_namespace(a, b)
     a, b = xp_broadcast_promote(a, b, ensure_writeable=True, force_floating=True, xp=xp)
+    a = xp_atleast_nd(a, ndim=1)
+    b = xp_atleast_nd(b, ndim=1) if b is not None else b
     axis = tuple(range(a.ndim)) if axis is None else axis
 
-    if b is not None:
-        a[b == 0] = -xp.inf
-
-    # Scale by real part for complex inputs, because this affects
-    # the magnitude of the exponential.
-    if xp_size(a) == 0:
-        # because `xp.max` doesn't have `initial` argument...
-        shape = np.asarray(a.shape)  # NumPy is concise for scalar or tuple `axis`
+    if xp_size(a) != 0:
+        with np.errstate(divide='ignore', invalid='ignore'):  # log of zero is OK
+            out, sgn = _logsumexp(a, b, axis=axis, return_sign=return_sign, xp=xp)
+    else:
+        shape = np.asarray(a.shape)  # NumPy is convenient for shape manipulation
         shape[axis] = 1
-        a_max = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        out = xp.full(tuple(shape), -xp.inf, dtype=a.dtype)
+        sgn = xp.sign(out)
+
+    # Deal with shape details - reducing dimensions and convert 0-D to scalar for NumPy
+    out = xp.squeeze(out, axis=axis) if not keepdims else out
+    sgn = xp.squeeze(sgn, axis=axis) if (sgn is not None and not keepdims) else sgn
+    out = out[()] if out.ndim == 0 else out
+    sgn = sgn[()] if (sgn is not None and sgn.ndim == 0) else sgn
+
+    return (out, sgn) if return_sign else out
+
+
+def _elements_and_indices_with_max_real(a, axis=-1, xp=None):
```

Comment:
Would you prefer that this lives in this file rather than `_lib` for now?

Reply:
Yes, I took it out of the function to avoid obscuring the logsumexp algorithm, not for reuse by others.